docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,473 @@
+  Rust, even if iOS does move to the new Sync Manager.
+
+* For Desktop, we hope to start by reusing the existing "sync manager"
+  implemented by Desktop, and start moving individual engines across.
+
+* There may be some cross-crate issues even for the Rust implemented engines.
+  Or more specifically, we'd like to avoid assuming any particular linkage or
+  packaging of Rust implemented engines.
+
+Even with these complications, we expect the high-level component
+to be written in a platform specific language (ie, Kotlin or Swift) and our
+lower-level component to be implemented in Rust and delivered as part of the
+application-services library - but that's not a free-pass.
+
+# The responsibilities of the Sync Manager.
+
+The primary responsibilities of the "high level" portion of the sync manager are:
+
+* Manage all FxA interaction. The low-level component will have a way to
+  communicate auth related problems, but it is the high-level component
+  which takes concrete FxA action.
+
+* Expose all UI for the user to choose what to sync and coordinate this with
+  the low-level component. Note that because these choices can be made on any
+  connected device, these choices must be communicated in both directions.
+
+* Implement timers or any other mechanism to fully implement the "sync
+  scheduler", including any policy decisions such as only syncing on WiFi,
+  etc.
+
+* Implement a UI so the user can "sync now".
+
+* Collect telemetry from the low-level component, probably augment it, then
+  submit it to the telemetry pipeline.
+
+The primary responsibilities of the "low level" portion of the sync manager are:
+
+* Manage the `meta/global`, `crypto/keys` and `info/collections` resources,
+  and interact with each engine as necessary based on the content of these
+  resources.
+
+* Manage interaction with the token server.
+
+* Enforce constraints necessary to ensure the entire ecosystem is not
+  subject to undue load. For example, this component should ignore attempts to
+  sync continiously, or to sync when the services have requested backoff.
+
+* Manage the "clients" collection - we probably can't ignore this any longer,
+  especially for bookmarks (as desktop will send a wipe command on bookmark
+  restore, and things will "be bad" if we don't see that command).
+
+* Define a minimal "service state" so certain things can be coordinated with
+  the high-level component. Examples of these states are "everything seems ok",
+  "the service requested we backoff for some period", "an authentication error
+  occurred", and possibly others.
+
+* Perform, or coordinate, the actual sync of the engines - from the containing
+  app's POV, there's a single "sync now" entry-point (in practice there might
+  be a couple, but conceptually there's a single way to sync). Note that as
+  mentioned above, this may involve coordinating with engines which aren't
+  implemented in Rust.
```

Comment:
Does it make sense to push this into the high-level manager? We talked a bit about this yesterday, but it seems like trying to expose Rust thunks for existing Swift or JS engines is going to be tricky. Since Firefox for iOS and Desktop already have something like a sync manager, what if we had the high-level manager take care of the existing engines, and also call into the low-level Rust manager for Rust engines?

Reply:
> Does it make sense to push this into the high-level manager?

Yeah, I agree it possibly does (although the low-level component will probably still need some knowledge for info/collections etc). I think we can see what shape we end up with before deciding this, so I added a new section "stuff tbd" and added this there.